### PR TITLE
HOSTEDCP-1709: test/e2e: use the eventually abstraction in chaos test

### PR DESCRIPTION
**Pull request by stevekuznetsov**

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Lots of copy-pasta in this file ... I can address it if you'd like in this PR, figured I'd err on the side of being minimal at first.

File: test/e2e/chaos_test.go [hunk 1]
```diff
@@ -71,22 +74,14 @@ func testKillRandomMembers(parentCtx context.Context, client crclient.Client, cl
...
-				}
-				return false, nil
-			}
-			return true, nil
-		})
-		g.Expect(err).NotTo(HaveOccurred(), "failed to create marker configmap")
+		e2eutil.EventuallyObject(t, ctx, "create marker", func(ctx context.Context) (*corev1.ConfigMap, error) {
+			err := guestClient.Create(ctx, cm)
+			return cm, err
+		}, e2eutil.NoOpPredicate[*corev1.ConfigMap]())
```
Comment (stevekuznetsov) [inline test/e2e/chaos_test.go:80]: I'm not thrilled at how users need to explicitly say they have no predicate, open to ideas here. We already have options pattern - perhaps we have flavors of this func that make the distinction explicit in their name?

File: test/e2e/chaos_test.go [hunk 2]
```diff
@@ -118,38 +113,25 @@ func testKillRandomMembers(parentCtx context.Context, client crclient.Client, cl
...
-					previousSTSError = err.Error()
-				}
-				return false, nil
-			}
-			return *etcdSts.Spec.Replicas == etcdSts.Status.ReadyReplicas, nil
-		})
-		g.Expect(err).NotTo(HaveOccurred(), "etcd statefulset available replicas never converged")
-		t.Logf("etcd statefulset recovered successfully")
+		e2eutil.EventuallyObject(t, ctx, "etcd StatefulSet replicas to converge", func(ctx context.Context) (*appsv1.StatefulSet, error) {
+			sts := cpomanifests.EtcdStatefulSet(guestNamespace)
```
Comment (stevekuznetsov) [inline test/e2e/chaos_test.go:117]: n.b. it's likely good hygiene to have individual variables + spots on the heap where we keep these objects. The current approach that over-writes e.g. the `etcdSts` var during the entire test's runtime is harder to follow in my opinion, often the value of that var depends on the control flow that got you to a particular spot in the test

File: test/e2e/chaos_test.go [hunk 3]
```diff
@@ -309,6 +266,7 @@ func testSingleMemberRecovery(parentCtx context.Context, client crclient.Client,
 			Namespace:     manifests.HostedControlPlaneNamespace(cluster.Namespace, cluster.Name),
 			LabelSelector: labels.Set(etcdSts.Spec.Selector.MatchLabels).AsSelector(),
 		})
+		g.Expect(err).NotTo(HaveOccurred(), "failed to get etcd pods")
```
Comment (stevekuznetsov) [inline test/e2e/chaos_test.go:269]: I am strongly biased against Ginkgo/Gomega, but I will note that I have never seen so many missed error checks as in a codebase leaning on Gomega for assertions!

**Comment by enxebre** [comment]
/lgtm
can you please share diff of the e2e output before / after this?

**Comment by stevekuznetsov** [comment]
The diffs should be minimal, since I already changed the existing code from what it used  to be, but anyway:

```
    chaos_test.go:305: waiting for etcd StatefulSet replicas to converge
    eventually.go:115: observed *v1.StatefulSet e2e-clusters-wvq4b-example-x2bml/etcd invalid at RV 17149 after 23.96005ms:
    eventually.go:117:  - wanted 3 replicas in spec, got 2 in status
    eventually.go:115: observed *v1.StatefulSet e2e-clusters-wvq4b-example-x2bml/etcd valid at RV 18596 after 4.988471773s:
    eventually.go:117:  - wanted 3 replicas in spec, got 3 in status
    chaos_test.go:305: Successfully waited for etcd StatefulSet replicas to converge in 20s
```

FWIW @sjenning with the 5s poll we're seeing this impact the test runtime already - more than one async assertion takes just 5s to complete, and even in this case it could have increased the time by 5s from 15 to 20, which is 33% longer. To whatever extent we want to cut down on e2e time, I think aggressive polls may be a good idea until we can show they actually have deleterious effects on the KAS.

**Comment by sjenning** [comment]
/hold

needs a rebase

```
test/e2e/chaos_test.go:80:14: undefined: e2eutil.NoOpPredicate
test/e2e/chaos_test.go:120:6: type func(sts *"k8s.io/api/apps/v1".StatefulSet) (done bool, reasons []string, err error) of func(sts *appsv1.StatefulSet) (done bool, reasons []string, err error) {…} does not match inferred type []"github.com/openshift/hypershift/test/e2e/util".Predicate[*"k8s.io/api/apps/v1".StatefulSet] for []"github.com/openshift/hypershift/test/e2e/util".Predicate[T]
test/e2e/chaos_test.go:131:6: type func(configMap *"k8s.io/api/core/v1".ConfigMap) (done bool, reasons []string, err error) of func(configMap *corev1.ConfigMap) (done bool, reasons []string, err error) {…} does not match inferred type []"github.com/openshift/hypershift/test/e2e/util".Predicate[*"k8s.io/api/core/v1".ConfigMap] for []"github.com/openshift/hypershift/test/e2e/util".Predicate[T]
test/e2e/chaos_test.go:165:14: undefined: e2eutil.NoOpPredicate
test/e2e/chaos_test.go:213:14: undefined: e2eutil.NoOpPredicate
test/e2e/chaos_test.go:227:6: type func(sts *"k8s.io/api/apps/v1".StatefulSet) (done bool, reasons []string, err error) of func(sts *appsv1.StatefulSet) (done bool, reasons []string, err error) {…} does not match inferred type []"github.com/openshift/hypershift/test/e2e/util".Predicate[*"k8s.io/api/apps/v1".StatefulSet] for []"github.com/openshift/hypershift/test/e2e/util".Predicate[T]
test/e2e/chaos_test.go:238:6: type func(configMap *"k8s.io/api/core/v1".ConfigMap) (done bool, reasons []string, err error) of func(configMap *corev1.ConfigMap) (done bool, reasons []string, err error) {…} does not match inferred type []"github.com/openshift/hypershift/test/e2e/util".Predicate[*"k8s.io/api/core/v1".ConfigMap] for []"github.com/openshift/hypershift/test/e2e/util".Predicate[T]
test/e2e/chaos_test.go:300:6: type func(pod *"k8s.io/api/core/v1".Pod) (done bool, reasons []string, err error) of func(pod *corev1.Pod) (done bool, reasons []string, err error) {…} does not match inferred type []"github.com/openshift/hypershift/test/e2e/util".Predicate[*"k8s.io/api/core/v1".Pod] for []"github.com/openshift/hypershift/test/e2e/util".Predicate[T]
test/e2e/chaos_test.go:309:6: type func(sts *"k8s.io/api/apps/v1".StatefulSet) (done bool, reasons []string, err error) of func(sts *appsv1.StatefulSet) (done bool, reasons []string, err error) {…} does not match inferred type []"github.com/openshift/hypershift/test/e2e/util".Predicate[*"k8s.io/api/apps/v1".StatefulSet] for []"github.com/openshift/hypershift/test/e2e/util".Predicate[T]
```

**Pull request merged**